File: ios/Classes/VPNChecker.swift
//
//  VPNManger.swift
//  vpn_check
//
//  Created by Denis Dudka on 11.01.2021.
//

import Foundation

struct VPNChecker {

    private static let vpnProtocolsKeysIdentifiers = [
        "tap", "tun", "ppp", "ipsec", "utun"
    ]

    static func isVpnActive() -> Bool {
        guard let cfDict = CFNetworkCopySystemProxySettings() else { return false }
        let nsDict = cfDict.takeRetainedValue() as NSDictionary
        guard let keys = nsDict["__SCOPED__"] as? NSDictionary,
            let allKeys = keys.allKeys as? [String] else { return false }

        // Checking for tunneling protocols in the keys
        for key in allKeys {
            for protocolId in vpnProtocolsKeysIdentifiers
                where key.starts(with: protocolId) {
                return true
            }
        }
        return false
    }
}

File: android/src/main/kotlin/com/cr0manty/vpn_check/VPNChecker.kt
package com.cr0manty.vpn_check

import java.net.NetworkInterface
import java.util.*

class VPNChecker {
    companion object {
        fun isVpnActive(): Boolean {
            val vpnProtocolsKeysIdentifiers = arrayOf("tap", "tun", "ppp", "ipsec", "utun")
            try {
                for (networkInterface in Collections.list(NetworkInterface.getNetworkInterfaces())) {
                    for (protocolId in vpnProtocolsKeysIdentifiers) {
                        if (networkInterface.isUp && networkInterface.name.startsWith(protocolId)) return true
                    }
                }
            } catch (ex: Exception) {
                return false
            }
            return false
        }
    }
}